### PR TITLE
WIP: Omniauth strategy

### PR DESCRIPTION
**Pull request by DanHenton**

### CODE REVIEW

File: lib/omniauth/strategies/realme.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+module OmniAuth
+  module Strategies
+    class Realme
+      include OmniAuth::Strategy
+      autoload :AuthRequest,  'omniauth/strategies/realme/auth_request'
+      autoload :AuthResponse, 'omniauth/strategies/realme/auth_response'
+
+      # Fixed OmniAuth options
+      option :provider, 'realme'
```
Comment (DanHenton) [inline lib/omniauth/strategies/realme.rb:13]: This has a different context to the omniauth provider. So is not a fixed option

File: lib/omniauth/strategies/realme.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+      def request
+        OmniAuth::Strategies::Realme::AuthRequest.new(self.class.default_options).call
+      end
+
+      def callback_phase
```
Comment (DanHenton) [inline lib/omniauth/strategies/realme.rb:20]: I don't think this will be needed as it should be handled by devise

File: spec/support/realme.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+OmniAuth::Strategies::Realme.configure do |config|
```
Comment (DanHenton) [inline spec/support/realme.rb]: This will probably need to be updated with mts configuration

File: lib/omniauth/strategies/realme.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+      def saml_settings
+        idp_metadata_parser = OneLogin::RubySaml::IdpMetadataParser.new
+        settings = idp_metadata_parser.parse(File.read(options.fetch('idp_service_metadata')))
+
+          settings.issuer                         = options.fetch('issuer')
+          settings.assertion_consumer_service_url = options.fetch('assertion_consumer_service_url')
+          settings.private_key                    = options.fetch('private_key')
+
+          settings.authn_context = 'urn:nzl:govt:ict:stds:authn:deployment:GLS:SAML:2.0:ac:classes:LowStrength' # TODO get from config
```
Comment (DanHenton) [inline lib/omniauth/strategies/realme.rb]: configuration

File: spec/support/realme.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
```
Comment (DanHenton) [inline spec/support/realme.rb]: update or remove. We

File: lib/omniauth/strategies/realme.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    class Realme
+      include OmniAuth::Strategy
+      autoload :AuthRequest, 'omniauth/strategies/realme/auth_request'
+
+      # Fixed OmniAuth options
+      option :provider, 'realme'
+
+      def request_phase
+        req = OneLogin::RubySaml::Authrequest.new
+        redirect req.create(saml_settings, 'SigAlg' => 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256')
```
Comment (yarhtut) [inline lib/omniauth/strategies/realme.rb:17]: `SigAlg` wrong spelling?

File: lib/omniauth/strategies/realme.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        super
+      end
+
+      private
+
+      def saml_settings
+        idp_metadata_parser = OneLogin::RubySaml::IdpMetadataParser.new
+        settings = idp_metadata_parser.parse(File.read(options.fetch('idp_service_metadata')))
+
+          settings.issuer                         = options.fetch('issuer')
```
Comment (yarhtut) [inline lib/omniauth/strategies/realme.rb:39]: may be indentation is off.

**Review by yarhtut — APPROVED** [review]
👍

File: README.md [hunk 8]
```diff
@@ -1,2 +1,159 @@
...
+  - You will need to update your registration `new` and `edit` views by adding the new fields as well as hidden fields for `provider` and `uid`.
+  - User sign in view will also need to be updated so that it links to the OmniAuth realme pass through using the link helper `user_realme_omniauth_authorize_path`.
+
+Model
+```ruby
+# app/models/user.rb
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
```
Comment (invke) [inline README.md:105]: From my understanding (totally may be wrong), I don't think they'd want `:database_authenticatable` - unless they were wanting to auth with a password against the DB as an alternative to the RealMe SAML auth.

**Review by invke — APPROVED** [review]
Seems all good man, good work! Nice README also 👍

File: README.md (hunk 8 above)
Comment (invke) [reply README.md:105]: Just poked my head into devise quickly, here's the related model data and their associated modules (copied from src):
```
      ## Database authenticatable
      t.string :email,              null: false, default: ""
      t.string :encrypted_password, null: false, default: ""

      ## Recoverable
      t.string   :reset_password_token
      t.datetime :reset_password_sent_at

      ## Rememberable
      t.datetime :remember_created_at

      ## Trackable
      # t.integer  :sign_in_count, default: 0, null: false
      # t.datetime :current_sign_in_at
      # t.datetime :last_sign_in_at
      # t.#{ip_column} :current_sign_in_ip
      # t.#{ip_column} :last_sign_in_ip

      ## Confirmable
      # t.string   :confirmation_token
      # t.datetime :confirmed_at
      # t.datetime :confirmation_sent_at
      # t.string   :unconfirmed_email # Only if using reconfirmable

      ## Lockable
      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
      # t.string   :unlock_token # Only if unlock strategy is :email or :both
      # t.datetime :locked_at
```

Comment (DanHenton) [reply README.md:105]: I would still suggest having `:database_authenticatable` as the required configuration option. We still need it inside our application for various authentication reasons. For Example Realme, ITE and production require a fixed domain to access accounts. 

By only allowing Realme devise logins developers would require annoying amounts of setup just to sign in a user correctly, for staging and development. 

ITE changes require Datacom and MTS requires `xml` file changes to point to a ngrok url, `Ngrok` which also needs to be setup, as well as the saml Issuer needs to be updated.

Developers then also need to update application configuration changes to get the `SAML` configuration correct.

I also have some security concerns about **not** assigning passwords to all accounts even if they never use it to log in.

File: lib/omniauth/strategies/realme.rb (hunk 6 above)
Comment (DanHenton) [reply lib/omniauth/strategies/realme.rb:17]: `SigAlg` is correct.

**Pull request merged**